src/hooks/mutations/useApprove.ts

```diff
@@ -0,0 +1,102 @@
+import { useAllowance } from 'react-queries/hooks/useAllowance'
+import type { MaybeApproveInput } from 'react-queries/queries/mutations'
+import { encodeFunctionData, getAddress } from 'viem'
+import { useEvmFees } from 'hooks/queries/useEvmFees'
+import { useWallet } from 'hooks/useWallet/useWallet'
+import { bnOrZero } from 'lib/bignumber/bignumber'
+import { isToken } from 'lib/utils'
+
+type UseApproveProps = MaybeApproveInput & {
+  onSuccess?: (txId: string) => void
+}
+
+export const useApprove = ({ onSuccess: handleSuccess, ...input }: UseApproveProps) => {
+  const queryClient = useQueryClient()
+  const wallet = useWallet().state.wallet
+
+  const [approvalTxId, setApprovalTxId] = useState<string | null>(null)
+
+  const chainId = useMemo(
+    () => (input.assetId ? fromAssetId(input.assetId).chainId : undefined),
+    [input.assetId],
+  )
+
+  const approvalCallData = useMemo(() => {
+    if (!(input.spender && input.amountCryptoBaseUnit)) return
+
+    return encodeFunctionData({
+      abi: erc20ABI,
+      functionName: 'approve',
+      args: [getAddress(input.spender), BigInt(input.amountCryptoBaseUnit)],
+    })
+  }, [input.amountCryptoBaseUnit, input.spender])
+
+  const maybeInputWithWallet = useMemo(
+    () => ({ ...input, wallet: wallet ?? undefined, data: approvalCallData }),
+    [approvalCallData, input, wallet],
+  )
+
+  const approvalFeesQueryInput = useMemo(
+    () => ({
+      value: '0',
+      accountNumber: input.accountNumber,
+      to: input.assetId ? fromAssetId(input.assetId).assetReference : undefined,
+      data: approvalCallData,
+      chainId,
+    }),
+    [approvalCallData, chainId, input.accountNumber, input.assetId],
+  )
+
+  const allowanceDataQuery = useAllowance({
+    assetId: input.assetId,
+    spender: input.spender,
+    from: input.from,
+  })
+
+  const isApprovalRequired = useMemo(() => {
+    if (!(input.assetId && input.amountCryptoBaseUnit && chainId)) return false
+    if (!(isEvmChainId(chainId) && isToken(fromAssetId(input.assetId).assetReference))) return false
+
+    if (!allowanceDataQuery?.data) return
+
+    const allowanceCryptoBaseUnit = allowanceDataQuery.data
+    return bnOrZero(input.amountCryptoBaseUnit).gt(allowanceCryptoBaseUnit)
+  }, [allowanceDataQuery.data, chainId, input.amountCryptoBaseUnit, input.assetId])
+
+  const approvalFeesQuery = useEvmFees({
+    ...approvalFeesQueryInput,
+    enabled: Boolean(isApprovalRequired),
+    staleTime: 30_000,
+    // Ensures fees are refetched at an interval, including when the app is in the background
+    refetchIntervalInBackground: true,
+    // Yeah this is arbitrary but come on, Arb is cheap
+    refetchInterval: isApprovalRequired ? 15_000 : false,
+  })
+
+  const approveMutation = useMutation({
+    ...reactQueries.mutations.approve(maybeInputWithWallet),
+    onSuccess: (txId: string) => {
+      handleSuccess?.(txId)
+      if (!(input.assetId && input.spender && input.from)) return
```

Comment:
Just to satisfy the type-system - we won't end up in here if these are undefined. No `skipToken` support for mutations, but these are asserted in the `approve` mutation.